Spigot-Server-Patches/0349-Use-ConcurrentHashMap-in-JsonList.patch

```diff
@@ -0,0 +1,115 @@
+From a3ff7e5e2422aaaee4d68c90397201107ef605f6 Mon Sep 17 00:00:00 2001
+From: egg82 <phantom_zero@ymail.com>
+Date: Tue, 7 Aug 2018 01:24:23 -0600
+Subject: [PATCH] Use ConcurrentHashMap in JsonList
+
+This is specifically aimed at fixing #471
+
+Using a ConcurrentHashMap because thread safety
+The performance benefit of Map over ConcurrentMap is negligabe at best in this scenaio, as most operations will be get and not add or remove
+Even without considering the use-case the benefits are still negligable
+
+Original ideas for the system included an expiration policy and/or handler
+The simpler solution was to use a computeIfPresent in the get method
+This will simultaneously have an O(1) lookup time and automatically expire any values
+Since the get method (nor other similar methods) don't seem to have a critical need to flush the map to disk at any of these points further processing is simply wasteful
+Meaning the original function expired values unrelated to the current value without actually having any explicit need to
+
+The h method was heavily modified to be much more efficient in its processing
+Also instead of being called on every get, it's now called just before a save
+This will eliminate stale values being flushed to disk
+
+Modified isEmpty to use the isEmpty() method instead of the slightly confusing size() < 1
+The point of this is readability, but does have a side-benefit of a small microptimization
+
+Finally, added a couple obfhelpers for the modified code
+
+diff --git a/src/main/java/net/minecraft/server/JsonList.java b/src/main/java/net/minecraft/server/JsonList.java
+index 0859c7eb2..05b3a1910 100644
+--- a/src/main/java/net/minecraft/server/JsonList.java
++++ b/src/main/java/net/minecraft/server/JsonList.java
+@@ -26,6 +26,8 @@ import java.util.Collection;
+ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Map;
++import java.util.concurrent.ConcurrentMap;
++
+ import org.apache.commons.io.IOUtils;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
+@@ -35,7 +37,8 @@ public class JsonList<K, V extends JsonListEntry<K>> {
+     protected static final Logger a = LogManager.getLogger();
+     protected final Gson b;
+     private final File c;
+-    private final Map<String, V> d = Maps.newHashMap();
++    // Paper - replace HashMap is ConcurrentHashMap
++    private final ConcurrentMap<String, V> d = Maps.newConcurrentMap(); private final ConcurrentMap<String, V> getBackingMap() { return this.d; } // Paper - OBFHELPER
```

Comment:
Is it really necessary to change the type of the field? Do you use any method that exists only on `ConcurrentMap` and not `Map`?

Reply:
The idea is that a later patch can improve further by checking this in, say, an AsyncPlayerPreLogin event (or method, etc)

Reply:
I'm not concerned about the use of `ConcurrentMap`, but since `ConcurrentMap` extends `Map`, you can just replace:

```java
private final Map<String, V> d = Maps.newHashMap();
```

with

```java
private final Map<String, V> d = Maps.newConcurrentMap();
```

without changing the type of the field.

Reply:
Ah, right. That's actually a very valid point. I suppose I don't see a use for any functions ConcurrentMap has anywhere else in that class. Possibly a change is in order, there. And an import removal.